### PR TITLE
Implicit polymorphism issue55

### PR DESCRIPTION
**Pull request by matthewbdwyer**
This is a pretty substantial change, but I'm hoping someone can look it over relatively quickly since the semester is fast approaching.

### CODE REVIEW

File: src/frontend/ast/ASTBuilder.cpp
Comment (will-leeson) [inline src/frontend/ast/ASTBuilder.cpp:1]: Are there any tests for poly? Would they cover this?

File: src/frontend/ast/treetypes/ASTFunction.cpp
Comment (will-leeson) [inline src/frontend/ast/treetypes/ASTFunction.cpp:1]: Same here

File: src/semantic/types/constraints/PolyTypeConstraintVisitor.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ * type with unique free type variables, i.e., the instantiated type,
+ * that can be unified with the function expression in the function application.
+ *
+ * An additional complexity is that we will perform this process for each
+ * function that may be called from this call site.  Since unification solves
+ * the conjunction of constraints this forces all of the corresponding free
+ * type variables for each called function to be consistently typed at this call
+ * site.  Different typings are possible at other call sites.
+ *
+ * FIX THE ABOVE ONCE THINGS ARE WORKING
```
Comment (will-leeson) [inline src/semantic/types/constraints/PolyTypeConstraintVisitor.cpp:29]: Should this line be removed?

**Review by will-leeson — COMMENTED** [review]
Besides the comments that needs to be removed, this looks ready.
One question, there seems to be somethings that should be covered using the --pi tests. I left comments on some. Should those have been covered?

File: src/frontend/ast/ASTBuilder.cpp
Comment (matthewbdwyer) [reply src/frontend/ast/ASTBuilder.cpp:1]: There are a couple of unit tests, e.g., PolyTypeConstraintCollectTest, and system tests, e.g., the new .../system/polytests.

Comment (matthewbdwyer) [reply src/frontend/ast/ASTBuilder.cpp:1]: There is another PR being resolved about the code coverage calculation.  Once that is fixed we can finalize 100\% coverage, but thanks for pointing this out.

File: src/frontend/ast/treetypes/ASTFunction.cpp
Comment (matthewbdwyer) [reply src/frontend/ast/treetypes/ASTFunction.cpp:1]: Same comment as above for ASTBuilder.cpp

File: src/semantic/types/constraints/PolyTypeConstraintVisitor.cpp (hunk 1 above)
Comment (matthewbdwyer) [reply src/semantic/types/constraints/PolyTypeConstraintVisitor.cpp:29]: Nice catch.  This required a substantial rewrite and I've done that now.

**Review by will-leeson — APPROVED** [review]

**Pull request merged**